### PR TITLE
Add minimalistic implementation of subgraph matcher.

### PR DESCRIPTION
**Pull request by ZolotukhinM**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#19322 Add minimalistic implementation of subgraph matcher.**

Differential Revision: [D14962182](https://our.internmc.facebook.com/intern/diff/D14962182)

### CODE REVIEW

**Comment by ZolotukhinM** [comment]
We can later add more features to this (e.g. nice iterators as in #17100). I'm putting this version for review now to unblock some of other work items.

File: torch/csrc/jit/subgraph_matcher.cpp [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    if (!matchValues(n1->inputs()[i], n2->inputs()[i])) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+/**
+ * Recursively try to match pattern with the actual graph starting from the
```
Comment (bwasti) [inline torch/csrc/jit/subgraph_matcher.cpp]: is this recursive?

File: torch/csrc/jit/subgraph_matcher.cpp [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+  std::vector<Match> matches;
+  std::stack<const Block*> blocks_to_visit;
+
+  // Iterate over all nodes in the graph (including nodes in subblocks) trying
+  // to match the pattern each node.
+  blocks_to_visit.push(graph.block());
+  while (!blocks_to_visit.empty()) {
+    const Block* block = blocks_to_visit.top();
+    blocks_to_visit.pop();
+    for (const Node* n : block->nodes()) {
```
Comment (bwasti) [inline torch/csrc/jit/subgraph_matcher.cpp]: maybe call out that this is a match on the imperative interpretation of the IR, and not the data flow graph

diamonds declared in different orders won't match, right?

**Review by bwasti — APPROVED** [review]
looks very good -- well tested

File: torch/csrc/jit/subgraph_matcher.cpp (hunk 1 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/subgraph_matcher.cpp]: > is this recursive?

Yes, `matchNodes` call `matchValues`, and `matchValues` call `matchNodes`. The recursion stops when we reach `Param` nodes in the pattern or if we hit an already visited node/value.

File: torch/csrc/jit/subgraph_matcher.cpp (hunk 2 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/subgraph_matcher.cpp]: Hm, they should match whatever the order is (I added a test for this). We do follow data flow graph, not the sequential list of nodes: `matchNodes` call `matchValues` for nodes' inputs, and `matchValues` proceed to values' defs calling `matchNodes` for them.

PS: This implementation lacks aliasing checks, which are required if we're going to replace matched subgraphs with something else - I don't forget about it, just plan to follow up on that separately.

**Review by ZolotukhinM — COMMENTED** [review]
Thanks for taking a look! I replied inline.

**Pull request closed without merging**